jlm/util/BiMap.hpp

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright 2023 Håvard Krogstie <krogstie.havard@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_UTIL_BI_MAP_HPP
+#define JLM_UTIL_BI_MAP_HPP
+
+#include <jlm/util/common.hpp>
+
+#include <unordered_map>
+
+namespace jlm::util
+{
+/**
+ * Represents a bijective mapping between keys and values, with fast loopups by both key and value.
+ * Keys and values must be unique among themselves.
+ * @tparam K the type of keys
+ * @tparam V the type of values
+ */
+template<typename K, typename V>
+class BiMap
+{
+  /**
+   * Adds a key/value pair to only the reverse map.
+   * @throws jlm::util::error if the value already exists in the BiMap.
+   */
+  void
+  AddPairToReverse(K key, V value)
+  {
+    auto it = ReverseMap_.find(value);
+    if (it != ReverseMap_.end())
+      throw jlm::util::error("Value already exsists in BiMap");
+    ReverseMap_.insert({ value, key });
+  }
+
+public:
+  ~BiMap() noexcept = default;
+
+  BiMap() = default;
+
+  BiMap(const BiMap & other) = default;
+
+  BiMap(BiMap && other) noexcept = default;
+
+  BiMap &
+  operator=(const BiMap & other) = default;
+
+  BiMap &
+  operator=(BiMap && other) = default;
+
+  /**
+   * Constructs a BiMap from an existing unordered_map.
+   * @throws jlm::util::error if the values are not unique
+   */
+  explicit BiMap(const std::unordered_map<K, V> & forwardMap)
+      : ForwardMap_(forwardMap),
+        ReverseMap_()
+  {
+    for (auto & [key, value] : ForwardMap_)
+      AddPairToReverse(key, value);
+  }
+
+  /**
+   * Removes all key/value mappings from the BiMap.
+   */
+  void
+  Clear() noexcept
+  {
+    ForwardMap_.clear();
+    ReverseMap_.clear();
+  }
+
+  /**
+   * Gets the number of key/value mappings in the BiMap.
+   */
+  [[nodiscard]] std::size_t
+  Size() const noexcept
+  {
+    return ForwardMap_.size();
+  }
+
+  /**
+   * Inserts a new key/value pair into the BiMap,
+   * @param key the key to be inserted, with the associated value
+   * @param value the value to be inserted, associated with the key
+   * @throws jlm::util::error if the key or value already exists in the BiMap
+   */
+  void
```

Comment:
Alternative suggestion for signature: `bool Insert(const K & key, const V & value)`, where the bool indicates whether the value was inserted or not. The user can then decide themselves whether an exception would be appropriate. Maybe you could also provide both Insert methods?